### DISCUSSION TITLE
Enhancement Suggestion - Conditional Supply Paths

### DISCUSSION THREAD

**Discussion by govindsmenokee** [frePPLe/frepple#375] [Q&A]
Hi,
There are some cases where the supply path will change based on the available items. To Explain this properly let us take the case of automotive headlamp manufacturing where BOM contains LEDs that have different bin codes. The bin code will not be known during the time of placing a purchase order. The bin codes are only known post dispatch or on arrival of the LEDs based on which the resistors are matched in series to get the standardised lumens output from the headlight assembly.

Just wanted to know if adding a new feature for these conditional supply paths would benefit the entire community or is there a better way of doing it.

Im adding an attachment just to have a more clear picture on the scenario
<img width="1077" alt="Screenshot 2021-07-08 at 1 31 18 AM" src="https://user-images.githubusercontent.com/2034781/124821423-50dcdf80-df8c-11eb-9d96-e1c28a4c3c35.png">
Here SMT Bin 1, SMT Bin 2 and SMT Bin 3 are alternate operations with the same priority. the problem here is while placing an order for the led it will be for the common led where the bin code is unknown and when we actually have an update on the bin code we will update the purchase order and generate the plan again at that point the system should pick the path with the matching bin code led. for example post shipment if we get to know that LED Bin 1 is shipped we will update the purchase order from LED ---> LED Bin 1 and regenerating the plan should consider that path. The best option would be to add one more search mode where material availability should be considered for picking the alternate supply path.

Hoping for some good thoughts and discussions on this.

Regards,
Govind

**Comment by jdetaeye** [comment]
>>The best option would be to add one more search mode where material availability should be considered for picking the alternate supply path.

You can already do something similar with the "minimum cost" search mode:
- set the cost in the itemsupplier table to 1000
- set the cost in the operation table to 1
- set the search mode to "minimum cost" in the operation table

As a result, any of the supply path that can use stock/confirmed supply (and thus doesn't need to generate extra purchase order) will have a much lower cost.

**Reply by govindsmenokee** [reply]
Hi,
Tried out this approach. If the LED is of the same price the final cost will be same in that case the planning engine is again trying to place a purchase order. Attaching a screenshot for reference
<img width="1671" alt="Screenshot 2021-07-08 at 2 56 16 PM" src="https://user-images.githubusercontent.com/2034781/124898338-b963a500-dffc-11eb-9125-f665e5e8129d.png">
As you can see there are two confirmed orders of 50 each but again system has proposed a new purchase order since the cost is the same (50,000)

**Reply by jdetaeye** [reply]
If one of the alternates covers the complete demand, the approach I described will work to select a good alternate.
But indeed, as you correctly point out, that approach will not work if the production needs to be split over multiple alternates.

**Reply by govindsmenokee** [reply]
Is there any possibility where we can add a step where it checks the number of shortages for each supply path and takes the one with the least number of shortages?

**Reply by jdetaeye** [reply]
>Is there any possibility where we can add a step ...

Frepple being open source, there is always a possibility. :-)    It's however not a very trivial change.

The planning logic for alternate operations is here:
https://github.com/frePPLe/frepple-enterprise/blob/27d3a1e47ce7b0571745c04ecdff28999fb68e84/src/solver/solveroperation.cpp#L1522

> ... where it checks the number of shortages for each supply path and takes the one with the least number of shortages?

Just going for the alternate path with the lowest number of shortages isn't a very generic approach: a) it ignores the quantity of the shortage, b) it doesn't split the requirement over multiple alternates (ie same problem as in your previous comment still exists).
A generic approach would require 2 passes: In a first pass you plan as much as possible with stock and confirmed supply across any alternate (ie first pass doesn't allow purchasing more material).  In the second pass you plan the remaining quantity while allowing to buy extra material.

**Reply by gibhsu** [reply]
Johan,

Not sure If I have a similar requirement that I spilt several inventory (buffers) to multiple operations based on the label (Normal or Hold WIP). The idea is the demand will be fulfilled by the inventory first then decide the new manufacturing orders need to be started. This is more making send for order fulfillment. Do I have any chance to model it in FrePPle?

**Reply by gibhsu** [reply]
According to your comments in this thread "A generic approach would require 2 passes: In a first pass you plan as much as possible with stock and confirmed supply across any alternate (ie first pass doesn't allow purchasing more material). In the second pass you plan the remaining quantity while allowing to buy extra material."
Do it means I need to customize the FrePPLE for 2 passes to generate manufacturing orders after pegged all available stocks?

**Reply by jdetaeye** [reply]
Gib,

I'd see the 2 pass approach as a core frepple enhancement.   The algorithm would perform this search internally to use up all existing stocks and WIP on all alternates.  Only when all existing supply is exhausted would we launch a new production batch.

In the planning tool I used in a previous life (and you which one that is :-) ) it was implemented exactly like this.
